python/descriptor.c

```diff
@@ -38,113 +42,919 @@
+      },
+      (void*)&upb_enumdef_lookupnamez,
+      (void*)&upb_enumvaldef_name,
+  };
+  PyUpb_DescriptorBase* self = (void*)_self;
+  return PyUpb_ByNameMap_New(&funcs, self->def, self->pool);
+}
+
+static PyObject* PyUpb_EnumDescriptor_GetValuesByNumber(PyObject* _self,
+                                                        void* closure) {
+  static PyUpb_ByNumberMap_Funcs funcs = {
+      {
+          (void*)&upb_enumdef_valuecount,
+          (void*)&upb_enumdef_value,
+          (void*)&PyUpb_EnumValueDescriptor_Get,
+      },
+      (void*)&upb_enumdef_lookupnum,
+      (void*)&upb_enumvaldef_number,
+  };
+  PyUpb_DescriptorBase* self = (void*)_self;
+  return PyUpb_ByNumberMap_New(&funcs, self->def, self->pool);
+}
+
+static PyObject* PyUpb_EnumDescriptor_GetContainingType(PyObject* _self,
+                                                        void* closure) {
+  PyUpb_DescriptorBase* self = (void*)_self;
+  return PyUpb_Descriptor_Get(upb_enumdef_containingtype(self->def));
+}
+
+static PyObject* PyUpb_EnumDescriptor_GetHasOptions(PyObject* _self,
+                                                    void* closure) {
+  PyUpb_DescriptorBase* self = (void*)_self;
+  return PyBool_FromLong(upb_enumdef_hasoptions(self->def));
+}
+
+static PyObject* PyUpb_EnumDescriptor_GetOptions(PyObject* _self,
+                                                 PyObject* args) {
+  PyUpb_DescriptorBase* self = (void*)_self;
+  return PyUpb_DescriptorBase_GetOptions(self, upb_enumdef_options(self->def),
+                                         &google_protobuf_EnumOptions_msginit,
+                                         "google.protobuf.EnumOptions");
+}
+
+static PyObject* PyUpb_EnumDescriptor_CopyToProto(PyObject* _self,
+                                                  PyObject* py_proto) {
+  return PyUpb_DescriptorBase_CopyToProto(
+      _self, (PyUpb_ToProto_Func*)&upb_EnumDef_ToProto,
+      &google_protobuf_EnumDescriptorProto_msginit, py_proto);
+}
+
+static PyGetSetDef PyUpb_EnumDescriptor_Getters[] = {
+    {"full_name", PyUpb_EnumDescriptor_GetFullName, NULL, "Full name"},
+    {"name", PyUpb_EnumDescriptor_GetName, NULL, "last name"},
+    {"file", PyUpb_EnumDescriptor_GetFile, NULL, "File descriptor"},
+    {"values", PyUpb_EnumDescriptor_GetValues, NULL, "values"},
+    {"values_by_name", PyUpb_EnumDescriptor_GetValuesByName, NULL,
+     "Enum values by name"},
+    {"values_by_number", PyUpb_EnumDescriptor_GetValuesByNumber, NULL,
+     "Enum values by number"},
+    {"containing_type", PyUpb_EnumDescriptor_GetContainingType, NULL,
+     "Containing type"},
+    {"has_options", PyUpb_EnumDescriptor_GetHasOptions, NULL, "Has Options"},
+    {NULL}};
+
+static PyMethodDef PyUpb_EnumDescriptor_Methods[] = {
+    {
+        "GetOptions",
+        PyUpb_EnumDescriptor_GetOptions,
+        METH_NOARGS,
+    },
+    {
+        "CopyToProto",
+        PyUpb_EnumDescriptor_CopyToProto,
+        METH_O,
+    },
+    {NULL}};
```

Comment:
why does this one have so much more whitespace?

Reply:
I think it was because there was a trailing comma. Fixed.